modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/PaginationBarTag.java

```diff
@@ -288,50 +301,135 @@ protected int processStartTag() throws Exception {
 
 			buttonTag.setCssClass("page-link");
 			buttonTag.setDisplayType("unstyled");
-			buttonTag.setLabel(String.valueOf(i));
+			buttonTag.setLabel(String.valueOf(firstItem));
 
 			buttonTag.doTag(pageContext);
 
-			jspWriter.write("</li>");
-		}
+			int leftBufferEnd = activeIndex - _ellipsisBuffer;
 
-		jspWriter.write("<li class=\"dropdown page-item\">");
+			int[] leftBuffer = Arrays.copyOfRange(
+				totalItems, 1, Math.max(leftBufferEnd, 1));
 
-		buttonTag = new ButtonTag();
+			if (leftBuffer.length > 1) {
+				jspWriter.write("<li class=\"dropdown page-item\">");
 
-		buttonTag.setCssClass("dropdown-toggle page-link");
-		buttonTag.setDisplayType("unstyled");
-		buttonTag.setLabel("...");
+				buttonTag = new ButtonTag();
 
-		buttonTag.doTag(pageContext);
+				buttonTag.setCssClass("dropdown-toggle page-link");
+				buttonTag.setDisplayType("unstyled");
+				buttonTag.setLabel("...");
 
-		jspWriter.write("</li>");
+				buttonTag.doTag(pageContext);
 
-		for (int i = ellipsisCount + 1; i < (totalPages + 1); i++) {
-			jspWriter.write("<li class=\"page-item");
+				jspWriter.write("</li>");
+			}
 
-			if (i == _activePage) {
-				jspWriter.write(" active");
+			int lastIndex = totalItems.length - 1;
+
+			int[] itemsToRender = Arrays.copyOfRange(
+				totalItems, Math.max(activeIndex - _ellipsisBuffer, 1),
+				Math.min(activeIndex + _ellipsisBuffer + 1, lastIndex));
+
+			for (int itemToRender : itemsToRender) {
+				jspWriter.write("<li class=\"page-item");
+
+				if (itemToRender == _activePage) {
+					jspWriter.write(" active");
+				}
+
+				if ((_disabledPages != null) &&
+					_disabledPages.contains(Integer.valueOf(itemToRender))) {
+
+					jspWriter.write(" disabled");
+				}
+
+				jspWriter.write("\">");
+
+				buttonTag = new ButtonTag();
+
+				buttonTag.setCssClass("page-link");
+				buttonTag.setDisplayType("unstyled");
+				buttonTag.setLabel(String.valueOf(itemToRender));
+
+				buttonTag.doTag(pageContext);
 			}
 
-			if ((_disabledPages != null) &&
-				_disabledPages.contains(Integer.valueOf(i))) {
+			int rightBufferStart = activeIndex + _ellipsisBuffer + 1;
 
-				jspWriter.write(" disabled");
+			if (rightBufferStart > lastIndex) {
+				rightBufferStart = lastIndex;
```

Comment:
@julien which case does a pagination need to be configured to fall into this use case? Maybe we also have to add in the React component?

Reply:
@matuzalemsteles I'll have to check because I'm not 100% but I think it was with these values.

```
activePage = 15
ellipsisBuffer = 2
totalItems = 17
```

About adding that to the react component, I'm not sure it's needed either, Java was throwing an `IndexOutOfBoundsException` but JavaScript (well TypeScript) doesn't even complain

Reply:
Hmm I see, no problem.